Description: Make kpartx wait until partitions are actually created.
 This should fix an issue with /dev/mapper/loop* partition not
 being available for mkfs step after kpartx as reported by many users.
 Extended to check the final umount call to close #740310
 .
 vmdebootstrap (0.2-2) unstable; urgency=medium
 .
   * Make kpartx wait whilst the partitions are created.
     (Closes: #741407)
   * Add a check on the cleanup umount and re-try if
     unable to complete. (Closes: #740310)

Author: Sunil Mohan Adapa <sunil@medhas.org>
Bug-Debian: http://bugs.debian.org/741407
Reviewed-By: Neil Williams <codehelp@debian.org>

Author: Neil Williams <codehelp@debian.org>
Bug-Debian: http://bugs.debian.org/740310
Reviewed-By: Neil Williams <codehelp@debian.org>

---

--- a/vmdebootstrap
+++ b/vmdebootstrap
@@ -23,6 +23,7 @@
 import shutil
 import subprocess
 import tempfile
+import time


 __version__ = '0.2'
@@ -220,7 +221,7 @@
         self.runcmd(['install-mbr', self.settings['image']])
 
     def setup_kpartx(self):
-        out = self.runcmd(['kpartx', '-av', self.settings['image']])
+        out = self.runcmd(['kpartx', '-avs', self.settings['image']])
         if self.settings['bootsize']:
             bootindex = 0
             rootindex = 1
@@ -469,10 +470,15 @@
         if self.settings['image']:
             for i in xrange(len(self.mount_points) - 1, -1, -1):
                 mount_point = self.mount_points[i]
-                self.runcmd(['umount', mount_point], ignore_fail=True)
+                try:
+                    self.runcmd(['umount', mount_point], ignore_fail=False)
+                except cliapp.AppException:
+                    logging.debug("umount failed, sleeping and trying again")
+                    time.sleep(5)
+                    self.runcmd(['umount', mount_point], ignore_fail=False)

             self.runcmd(['kpartx', '-d', self.settings['image']], ignore_fail=True)
-        
+
         for dirname in self.remove_dirs:
             shutil.rmtree(dirname)

